base/mpfr.jl

```diff
@@ -180,15 +180,14 @@ BigFloat(x, r::RoundingMode)
 widen(::Type{Float64}) = BigFloat
 widen(::Type{BigFloat}) = BigFloat
 
-function BigFloat(x::BigFloat, r::MPFRRoundingMode=ROUNDING_MODE[]; precision::Integer=DEFAULT_PRECISION[])
-    if precision == MPFR.precision(x)
-        return x
-    else
-        z = BigFloat(;precision=precision)
-        ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
-              z, x, r)
-        return z
-    end
+function BigFloat(x::BigFloat, r::MPFRRoundingMode=ROUNDING_MODE[];
+                    precision::Integer=MPFR.precision(x), new::Bool=false)
+    precision == MPFR.precision(x) && !new && return x
```

Comment:
```suggestion
    new || return x
```

Reply:
Does this sound good?
```julia
precision != MPFR.precision(x) && !new &&
                    @warn "new=false is ignored because precision differ" &&
                    return BigFloat(x, precision=precision)
new || return x
```

Reply:
The precision is being copied. In the constructor, default value of precision is precision(x).

Reply:
Ok, then with @simonbyrne's new default value I'm more ok with this.

Reply:
me too

Reply:
I recall a whole discussion where we concluded that `BigFloat(x::BigFloat)` should use the current global precision, not the precision of `x`. Are we sure we want to change that? It doesn't seem like something we can just change casually as part of dealing with `nextfloat` and copying.

The `new` keyword argument seems quite ad-hoc to me; since no other functions support it it's just a more verbose way to add a function for this purpose. It also didn't occur to me before that it conflicts with the precision behavior of the constructor, which I think is a pretty significant issue.

Reply:
I thought there was a later extended discussion where we concluded that using the global precision was most appropriate as a fallback, default and that ignoring the given precision of an extant `BigFloat` when copying it had been determined to be more problematic than beneficial.  If that is untrue, and I grant that it may be, then please note that much of my [advice](https://github.com/JuliaLang/julia/issues/31276) earlier was based upon that supposition.

Reply:
I was thinking of #29127. Is there a link to the later discussion?

Reply:
That's the one :blush:!

Reply:
I think we need a way to not force conformance to the global setting.  If I include a package that exports `const HiPrecisionPi` which is a `BigFloat` with a precision greater than the default, should it be incumbent on me to know that I must reset the global precision in order to work with, say,
`const HiPrecisionHalfPi = HiPrecisionPi/2` at the higher precision?

Reply:
I found a problem with this new BigFloat API. By using such implementation, `setprecision` function breaks. Consider the following example:
```julia
julia> x = BigFloat(12.12)
12.1199999999999992184029906638897955417633056640625

julia> setprecision(128) do
           y = BigFloat(x)
           y.prec
       end
256
```
I think this would cause a lot of problems as current behaviour of `setprecision` is:
```julia
julia> x = BigFloat(12.12)
12.1199999999999992184029906638897955417633056640625

julia> setprecision(128) do
           y = BigFloat(x)
           y.prec
       end
128
```

Reply:
Valid. What is your suggestion?

Reply:
The proper value of `nextfloat(::BigFloat)` or `prevfloat(::BigFloat)` is precision dependant.  The value obtained that defers to the global precision is not the same as the value obtained that defers to the precision of argument when that precision differs from the global precision.

It is reasonable to decide how we want to handle this.  If the determination is that `nextfloat` and `prevfloat` use the global precision then there is no reason to copy maintaining the precision of a the value copied [at least for this purpose].  If the determination is that `nextfloat` and `prevfloat` are requesting steps in units of the precision of the argument, then there is reason to have at least an internal function that copies maintaining the precision of the value copied.  In that case, we are saying something about the primacy of the precision attached to a given `BigFloat` over the global precision setting.  And that may warrant more adaption.

Reply:
This method was previously rejected but this might solve our problems:
```julia
function BigFloat(x::BigFloat, r::MPFRRoundingMode=ROUNDING_MODE[];
                           precision::Integer=0, new::Bool=false)
      precision==MPFR.precision(x) && !new && return x
      z = BigFloat(;precision=ifesle(precision==0,
                                  ifelse(new, MPFR.precision(x), DEFAULT_PRECISION[]), 
                                  precision))
      ccall((:mpfr_set, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, MPFRRoundingMode),
                z, x, r)
       return z
```
Maybe we can try to iterate over this method to achieve backward compatibilty.

Reply:
We should decide about using `new` as a parameter first. imo my prior note gives a way to determine that which strips away some of the dross.

Reply:
_trying to avoid going in circles, and attempting top recapitulate the generally favoured facets that have arisen while not involving any controversial aspects ..._

within mpfr.jl and __not__ exported:
these functions use the value and precision of `x`
```
function nextfloat!(x::BigFloat)
    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x) != 0
    return x
end

function prevfloat!(x::BigFloat)
    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x) != 0 
    return x
end
```

in mpfr.jl and exported
these functions respect the value and precision of `x`
```
nextfloat(x::BigFloat) = nextfloat!(x)
prevfloat(x::BigFloat) = prevfloat!(x)
```

`nextfloat!(x,n), nextfloat(x,n), prevfloat!(x,n), prevfloat(x,n)` accordingly

Reply:
I think this does what we want and avoids the controversial.

defined in mpfr and __not__ exported, these respect the precision of `x`
```
function nextfloat!(x::BigFloat)
    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ref{BigFloat},), x) != 0
    return x
end

function prevfloat!(x::BigFloat)
    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ref{BigFloat},), x) != 0 
   return x
end
```

defined in mpfr.jl and exported, these also respect the precision of `x`
without influencing any other aspects of the `BigFloat` interface.
```
nextfloat(x::BigFloat) = nextfloat!(x)
prevfloat(x::BigFloat) = prevfloat!(x)
```

`nextfloat(x,n), prevfloat(x,n), nextfloat!(x,n), prevfloat!(x,n)` accordingly

Reply:
^ revised

Reply:
+:100: to the non-controversial approach of introducing an unexported `duplicate` function for this PR. That let's us get this PR merged adding the desired `nextfloat` and `prevfloat` methods and we can have a separate discussion of the best API for creating copies.

Reply:
@narendrakpatel please ^, do as we discussed last night.